src/main/java/ch/njol/skript/expressions/ExprLore.java

```diff
@@ -1,236 +1,181 @@
-						}
-						addLore.add(line);
+				case SET, DELETE -> modifiedLore = (delta != null) ? List.of((String[]) delta) : null;
+				case ADD -> {
+					assert delta != null;
+					modifiedLore.addAll(List.of((String[]) delta));
+				}
+				case REMOVE, REMOVE_ALL -> {
+					boolean isAll = mode == ChangeMode.REMOVE_ALL;
+					if (isAll) {
+						//noinspection DataFlowIssue
+						modifiedLore.removeAll(List.of((String[]) delta));
+					} else {
+						//noinspection DataFlowIssue
+						for (String string : (String[]) delta)
+							modifiedLore.remove(string);
 					}
-					lore.addAll(addLore);
-					break;
-				case DELETE:
-					lore = null;
-					break;
-				case REMOVE:
-				case REMOVE_ALL:
-					assert stringDelta != null;
-					lore = Arrays.asList(handleRemove(
-							StringUtils.join(lore, "\n"), stringDelta[0], mode == ChangeMode.REMOVE_ALL).split("\n"));
-					break;
-				case RESET:
-					assert false;
-					return;
+				}
 			}
 		} else {
-			// Note: line number is changed from one-indexed to zero-indexed here
-			int lineNum = Math2.fit(0, lineNumber.intValue() - 1, 99); // TODO figure out the actual maximum
-
-			// Fill in the empty lines above the line being set with empty strings (avoids index out of bounds)
-			while (lore.size() <= lineNum)
-				lore.add("");
+			int loreIndex = this.lineNumber.getOptionalSingle(event).orElse(0).intValue() -1;
+			if (loreIndex < 0)
+				return; // Cannot change anything in lore if it's negative, therefor we return
 			switch (mode) {
-				case SET:
-					assert stringDelta != null;
-					lore.set(lineNum, stringDelta[0]);
-					break;
-				case ADD:
-					assert stringDelta != null;
-					lore.set(lineNum, lore.get(lineNum) + stringDelta[0]);
-					break;
-				case DELETE:
-					lore.remove(lineNum);
-					break;
-				case REMOVE:
-				case REMOVE_ALL:
-					assert stringDelta != null;
-					lore.set(lineNum, handleRemove(lore.get(lineNum), stringDelta[0], mode == ChangeMode.REMOVE_ALL));
-					break;
-				case RESET:
-					assert false;
-					return;
+				case SET -> {
+					for (int line = modifiedLore.size()-1; line < loreIndex; line++) {
+						modifiedLore.add("");
+					}
+					//noinspection DataFlowIssue
+					modifiedLore.set(loreIndex, (String) delta[0]);
+				}
+				case DELETE -> {
+					if (loreIndex > modifiedLore.size() || !itemMeta.hasLore())
+						return; // Cannot change anything in lore, therefor we return
+					modifiedLore.remove(loreIndex);
+				}
 			}
 		}
 
-		meta.setLore(lore == null || lore.size() == 0 ? null : lore);
-		if (i instanceof ItemStack) {
-			((ItemStack) i).setItemMeta(meta);
-		} else {
-			((ItemType) i).setItemMeta(meta);
+		if (modifiedLore != null && !modifiedLore.isEmpty()) {
+			if (IS_RUNNING_1_21_5) {
+				// The maximum amount of lore an item can have is 256
+				// this change was made in 1.21.5
+				// source: https://minecraft.wiki/w/Data_component_format#lore
+				modifiedLore = modifiedLore.stream().limit(256).toList();
```

Comment:
Is there an error if too much lore is passed?

Reply:
I'm honestly not sure, I didn't test it without the limitation as I just went off previous behavior. I'll keep a note for this to test it and see if we can remove it.

Reply:
Yep, I was thinking if we can remove it that would be good. If not, this is fine as-is.

Reply:
Didn't feel like loading a server to quickly test lore limit, it appears we're safer adding this paper's CraftMetaItem will throw an exception if invalid lore is provided 

https://github.com/PaperMC/Paper/blob/9b1798d6438107fdf0d5939b79a8cf71f4d16e2c/paper-server/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java#L1386-L1399

While I was here went and checked it doesn't seem like there was a previous lore limit prior to 1.21.5 internally however I don't want to take the chance there to remove it, so I believe we're safer keeping this implemented as in, I could maybe implement a static field for max lore instead of splitting it as I did here